### ISSUE TITLE
`rmarkdown::render` does not respect symlinks

### ISSUE THREAD

**Issue by SamDM** [rstudio/rmarkdown#1508]
When rendering a `.Rmd` file that is a symlink, the working directory becomes the parent directory of the original file where the symlink points to. As a result, the output appears in a different directory than the `.Rmd` file.

See the example below:
````
mkdir /tmp/foo
cd /tmp/foo
mkdir bar
mkdir baz
cat > bar/example.Rmd
---
title: Test
output: html_document
---

```{r}
getwd()
```
^D
ln -s ../bar/example.Rmd baz
````

The file structure looks like this:
```
/tmp/foo
├── bar
│   └── example.Rmd
└── baz
    └── example.Rmd -> ../bar/example.Rmd
```
Now I proceed to rendering the html:
```
cd baz
R -e "rmarkdown::render('example.Rmd',output_file='output.html')"
```
Now the file tree looks like this:
```
/tmp/foo
├── bar
│   ├── example.Rmd
│   └── output.html
└── baz
    └── example.Rmd -> ../bar/example.Rmd
```
Also, the output of `getwd()` in the html equals `"/tmp/foo/bar"`

In other words, running the command for rendering `/tmp/foo/baz/example.Rmd` resulted in `/tmp/foo/bar/example.Rmd` being rendered instead. That is, `render` followed the symlink before rendering the document. 

The expected behavior would be that `markdown::render` treats symlinks like normal files and does not follow them. At least, the current [docs](https://www.rdocumentation.org/packages/rmarkdown/versions/1.11/topics/render) do not mention anything about this special treatment for symlinks.

I'm not complaining though, `rmarkdown` is pretty neat. I just bumped into this and wanted to point it out :)

**Comment by earcanal** [comment]
I ran into this today, and agree that it should render into the same location as the symlink.

**Comment by yihui** [comment]
My guess is we called `normalizaPath()` somewhere in `render()` to resolve the symlink. I don't have time to dig into it. Sorry.

**Comment by jherrero** [comment]
Same here. In case this helps anyone, I got around by adding output_dir = "." as in:
R -e "rmarkdown::render('example.Rmd',output_dir='.',output_file='output.html')"

Not perfect, but it works.

**Comment by gorgitko** [comment]
@jherrero This still doesn't work - although `output.html` is created in the current working directory, the `example.Rmd` is executed in its source location (try to output `getwd()` in some chunk).

My current workaround is following:

```
rmd_tempfile <- tempfile(tmpdir = ".", fileext = ".Rmd")
file.copy("example.Rmd", rmd_tempfile)
rmarkdown::render(rmd_tempfile, output_dir = ".", output_file = "output.html")
file.remove(rmd_tempfile)
```

tl;dr Create a temporary copy of the source .Rmd in the current working directory.

**Comment by cderv** [comment]
@SamDM based on your example, as @yihui said, R will follow simlink when normalizing the PATH. 

```sh
$ cd /tmp/foo/baz
$ Rscript -e 'normalizePath("example.Rmd")'
[1] "/tmp/foo/bar/example.Rmd"
```

and we call it in `render()` on the input file : https://github.com/rstudio/rmarkdown/blob/69b54b68a586a59e0d8fc678e989e69eb8b3d376/R/render.R#L349

Also we execute the code in the input directory
https://github.com/rstudio/rmarkdown/blob/69b54b68a586a59e0d8fc678e989e69eb8b3d376/R/render.R#L385-L387

so it will also be in `foo/bar` not `foo/baz`

Setting the `output_dir="."` will output in the correct folder but the default working dir will be the one from the normalized path. `knit_root_dir` in `render` could solve that. 
```r
rmarkdown::render('example.Rmd',output_dir='.',output_file='output.html', knit_root_dir = normalizePath('.'))
```

it requires a `normalizePath` because it isn't normalized inside `render()`

So yes, this is an issue currently. **rmarkdown** does not work well with symlink currently. 

Maybe using **fs** to handle file paths in **rmarkdown** could help with all the paths manipulation. It would be the correct path if we want absolute one:
```sh
$ Rscript -e 'fs::path_abs("./example.Rmd")'
/tmp/foo/baz/example.Rmd
```

However, changing anything like this in **rmarkdown** could lead to side effects... 🤔

**Comment by yihui** [comment]
> The one-liner replacement is shown below:
> 
> ```r
> rmd <- file.path(normalizePath(dirname(rmd)), basename(rmd));
> ```

Yes, that's what I'd do.

> This is a naive question: What is a good way to support moving forward? Would I fork, update, test on my own? Then potentially file a pull request? I respect that people have other priorities. :)

Sorry for not getting back earlier. Now we have the PR #2438. Please feel free to test it. Thanks!

**Comment by jmw86069** [comment]
Wow this is great, thanks so much for your time and the updates! I will test as you described in the PR and comment accordingly.

**Comment by yihui** [comment]
@jmw86069 Have you had a chance to test it? Just asking. No hurry. Thanks!

**Comment by jmw86069** [comment]
Just to round out this issue, yes @yihui I have finally tested the update, and confirm that it fixes my scenario!

Thank you so much for the update.

I had been using a forked "rogue" version of rmarkdown that I edited for testing, and was quite nervous to submit a PR since it's an important package and pretty core change to its internals. One day I'll issue a PR... probably not this R package, but one day I promise. Haha.

**Comment by yihui** [comment]
@jmw86069 Totally understand. I was also a little nervous to make this change for the reason you mentioned. The new version of **rmarkdown** has been on CRAN for two weeks. We haven't received any bug reports yet, and I hope we won't :)
